fuse_optimizers/src/optimizer.cpp

```diff
@@ -100,36 +100,64 @@ void Optimizer::loadMotionModels()
     return;
   }
   // Validate the parameter server values
-  XmlRpc::XmlRpcValue motion_model_list;
-  private_node_handle_.getParam("motion_models", motion_model_list);
-  if (motion_model_list.getType() != XmlRpc::XmlRpcValue::TypeArray)
+  XmlRpc::XmlRpcValue motion_models;
+  private_node_handle_.getParam("motion_models", motion_models);
+  if (motion_models.getType() == XmlRpc::XmlRpcValue::TypeArray)
   {
-    throw std::invalid_argument("The 'motion_models' parameter should be a list of the form: "
-                                "-{name: string, type: string}");
+    // Validate all of the parameters before we attempt to create any plugin instances
+    for (int32_t motion_model_index = 0; motion_model_index < motion_models.size(); ++motion_model_index)
+    {
+      // Validate the parameter server values
+      if ( (motion_models[motion_model_index].getType() != XmlRpc::XmlRpcValue::TypeStruct)
+        || (!motion_models[motion_model_index].hasMember("name"))
+        || (!motion_models[motion_model_index].hasMember("type")))
+      {
+        throw std::invalid_argument("The 'motion_models' parameter should be a list of the form: "
+                                    "-{name: string, type: string}");
+      }
+    }
+    for (int32_t motion_model_index = 0; motion_model_index < motion_models.size(); ++motion_model_index)
+    {
+      // Get the setting we need from the parameter server
+      std::string motion_model_name = static_cast<std::string>(motion_models[motion_model_index]["name"]);
+      std::string motion_model_type = static_cast<std::string>(motion_models[motion_model_index]["type"]);
+      // Create a motion model object using pluginlib. This will throw if the plugin name is not found.
+      auto motion_model = motion_model_loader_.createUniqueInstance(motion_model_type);
+      // Initialize the publisher
+      motion_model->initialize(motion_model_name);
+      // Store the publisher in a member variable for use later
+      motion_models_.emplace(motion_model_name, std::move(motion_model));
+    }
```

Comment:
Can we separate out the parameter parsing from the plugin initialization?
Something like:
```
private_node_handle_.getParam("motion_models", motion_models);
struct PluginConfig
{
  std::string name;
  std::string type;
};
std::vector<PluginConfig> motion_model_configs;
if (motion_models.getType() == XmlRpc::XmlRpcValue::TypeArray)
{
if ( (motion_models[motion_model_index].getType() != XmlRpc::XmlRpcValue::TypeStruct)
    || (!motion_models[motion_model_index].hasMember("name"))
    || (!motion_models[motion_model_index].hasMember("type")))
  {
    throw std::invalid_argument("The 'motion_models' parameter should be a list of the form: "
                                "-{name: string, type: string}");
  }
  motion_model_configs.push_back(
    {static_cast<std::string>(motion_models[motion_model_index]["name"]),
     static_cast<std::string>(motion_models[motion_model_index]["type"])});
}
else if (motion_models.getType() == XmlRpc::XmlRpcValue::TypeStruct)
{
  if ( (motion_model_entry.second.getType() != XmlRpc::XmlRpcValue::TypeStruct)
    || (!motion_model_entry.second.hasMember("type")))
  {
    throw std::invalid_argument("The 'motion_models' parameter should be a struct of the form: "
                                "{string: {type: string}}");
  }
  motion_model_configs.push_back(
    {static_cast<std::string>(motion_model_entry.first),
     static_cast<std::string>(motion_model_entry.second["type"])});
}
else
{
  throw std::invalid_argument("The 'motion_models' parameter should be a list of the form: "
                              "-{name: string, type: string} or a struct of the form: {string: {type: string}}");
}
for (const auto& config : motion_model_configs)
{
  auto motion_model = motion_model_loader_.createUniqueInstance(config.type);
  // Initialize the motion model
  motion_model->initialize(config.name);
  // Store the motion model in a member variable for use later
  motion_models_.emplace(motion_model_name, std::move(motion_model));
}
```
That removes the repeated code from the two parameter server layout options. It's not that big of a deal for the motion models, but the sensor models are bit more involved.

Either that, or drop support for the list version. I'm fine with that as well.

Reply:
Done.

I prefer to keep backwards compatibility with the **list/array** format, at least for now. But it's probably a good idea to deprecate it, since the **struct** format looks superior.

I also added a test for the **array** configuration format, and organized the tests better.